plugin.video.vrt.nu/resources/lib/kodiwrapper.py

```diff
@@ -635,49 +623,53 @@ def refresh_caches(self, cache_file=None):
+        self.delete_file(self._cache_path + 'favorites.json')
         self.delete_file(self._cache_path + 'oneoff.json')
 
+    @staticmethod
+    def input_down():
+        ''' Move the cursor down '''
+        xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "Input.Down", "id": 1}')
+
     def container_refresh(self):
         ''' Refresh the current container '''
-        self.log_notice('Execute: Container.Refresh', 'Debug')
+        self.log('Execute: Container.Refresh', 'Debug')
         xbmc.executebuiltin('Container.Refresh')
 
-    def container_update(self, url=None, path='', replace=False):
-        ''' Update the current container '''
-        if url is None:
-            url = self._url
-        self.log_notice('Execute: Container.Update(%s%s%s)' % (url, path, ',replace' if replace else ''), 'Debug')
-        xbmc.executebuiltin('Container.Update(%s%s%s)' % (url, path, ',replace' if replace else ''))
-
     def end_of_directory(self):
         ''' Close a virtual directory, required to avoid a waiting Kodi '''
         xbmcplugin.endOfDirectory(handle=self._handle, succeeded=False, updateListing=False, cacheToDisc=False)
 
     def log_access(self, url, query_string=None, log_level='Verbose'):
         ''' Log addon access '''
-        if log_levels.get(log_level, 0) <= self._max_log_level:
+        if LOG_LEVELS.get(log_level, 0) <= self._max_log_level:
             message = '[%s] Access: %s' % (self._addon_id, url + ('?' + query_string if query_string else ''))
             xbmc.log(msg=from_unicode(message), level=xbmc.LOGNOTICE)
 
-    def log_notice(self, message, log_level='Info'):
+    def log(self, message, log_level='Info', **kwargs):
         ''' Log info messages to Kodi '''
-        if log_levels.get(log_level, 0) <= self._max_log_level:
-            message = '[%s] %s' % (self._addon_id, message)
-            xbmc.log(msg=from_unicode(message), level=xbmc.LOGNOTICE)
+        cur_log_level = LOG_LEVELS.get(log_level, 0)
+        if not self._debug_logging and 1 < cur_log_level <= self._max_log_level:
+            # If Debug Logging is not enabled, Kodi filters everything up to NOTICE out
+            log_level = 'Info'
+        if kwargs:
+            import string
+            message = string.Formatter().vformat(message, (), SafeDict(**kwargs))
+        message = '[{addon}] {message}'.format(addon=self._addon_id, message=message)
+        xbmc.log(msg=from_unicode(message), level=XBMC_LOG_LEVELS.get(log_level))
 
-    def log_error(self, message, log_level='Info'):
+    def log_error(self, message, **kwargs):
         ''' Log error messages to Kodi '''
-        message = '[%s] %s' % (self._addon_id, message)
+        if kwargs:
+            import string
+            message = string.Formatter().vformat(message, (), SafeDict(**kwargs))
+        message = '[{addon}] {message}'.format(addon=self._addon_id, message=message)
         xbmc.log(msg=from_unicode(message), level=xbmc.LOGERROR)
```

Comment:
Logging to the Kodi log is only allowed with the `DEBUG` level. See https://kodi.wiki/view/Add-on_rules:

> addons should use the xbmc.log() method to write to the logfile and use the debug logging level only.

Reply:
This is the case. But we make an exception for testing/development.